src/DictList.elm

```diff
@@ -937,6 +947,77 @@ fromDict dict =
     DictList dict (Dict.keys dict)
 
 
+{-| Takes a key-fn and a list.
+Creates a `Dict` which maps the key to a list of matching elements.
+    mary = {id=1, name="Mary"}
+    jack = {id=2, name="Jack"}
+    jill = {id=1, name="Jill"}
+    groupBy .id [mary, jack, jill] == DictList.fromList [(1, [mary, jill]), (2, [jack])]
+-}
+groupBy : (comparable1 -> a -> comparable2) -> DictList comparable1 a -> DictList comparable2 (List a)
+groupBy keyfn list =
+    foldr
+        (\key x acc ->
+            update (keyfn key x) (Maybe.map ((::) x) >> Maybe.withDefault [ x ] >> Just) acc
+        )
+        empty
+        list
+
+
+{-| Create a dictionary from a list of values, by passing a function that can get a key from any such value.
+If the function does not return unique keys, earlier values are discarded.
+This can, for instance, be useful when constructing Dicts from a List of records with `id` fields:
+    mary = {id=1, name="Mary"}
+    jack = {id=2, name="Jack"}
+    jill = {id=1, name="Jill"}
+    fromListBy .id [mary, jack, jill] == DictList.fromList [(1, jack), (2, jill)]
+-}
+fromListBy : (comparable1 -> a -> comparable2) -> DictList comparable1 a -> DictList comparable2 a
```

Comment:
Same comment as above -- I think this makes more sense as really being a way to construct a `DictList` from a `List`, not from another `DictList`. As written here, it's more of a variation on a `map` ... I suppose we could have another function and call it something like `mapKeys` if it's useful. Ah, except that as I read on we already have a `mapKeys` that just considers the key. I suppose we could have a `mapKeysAndValues`, in theory -- I haven't really thought it through.